theano/tensor/nnet/abstract_conv.py

```diff
@@ -357,13 +374,185 @@ def conv2d_grad_wrt_weights(input,
+    return gradWeight_op(input, output_grad, filter_shape[:-2])
+
+
+def bilinear_kernel_2D(ratio, normalize=True):
+    """ This function gets a ratio
+    and returns the normalized 2D filters that can
+    be applied to any image to upsample it by that ratio
+    using bilinear interpolation.
+
+    :type ratio: python integer or symbolic theano
+                 variable (Constant or Scalar)
+    : param ratio: the ratio by which an image will
+        be upsampled by the returned filters in the 2D space.
+
+    :type normalize: bool
+    : param normalize: indicates whether to normalize the
+    kernel or not. Default is True.
+
+    returns: the 2D kernels that can be applied to
+             any given image to upsample it by the indicated ratio
+             using bilinear interpolation.
+    """
+
+    hkern = bilinear_kernel_1D(ratio=ratio, normalize=normalize).dimshuffle('x', 0)
+    vkern = bilinear_kernel_1D(ratio=ratio, normalize=normalize).dimshuffle(0, 'x')
+    kern = hkern * vkern
+    return kern
+
+
+def bilinear_kernel_1D(ratio, normalize=True):
+    """ This function gets a ratio
+    and returns the normalized 1D filters that can
+    be applied to any image to upsample it by that ratio
+    using bilinear interpolation.
+
+    :type ratio: python integer or symbolic theano
+                 variable (Constant or Scalar)
+    : param ratio: the ratio by which an image is upsampled
+        in the 1D space.
+
+    :type normalize: bool
+    : param normalize: indicates whether to normalize the
+    kernel or not. Default is True.
+
+    returns: the 1D kernels that can be applied to
+             any given image to upsample it by the indicated ratio
+             using bilinear interpolation in one dimension.
+    """
+
+    T = theano.tensor
+    half_kern = T.arange(1, ratio + 1, dtype=theano.config.floatX)
+    kern = T.concatenate([half_kern, half_kern[-2::-1]])
+
+    if normalize:
+        kern /= float(ratio)
+    return kern
+
+
+def bilinear_upsampling(input,
+                        ratio,
+                        input_shape=None,
+                        use_1D_kernel=True):
+    """This function takes a stack of 2D inputs and upsamples it (in 2D space)
+    by the given ratio using bilinear interpolation with fixed (not learned)
+    kernels. When the upsampling ratio is even, the last row and colum is
+    repeated one extra time compared to the first row and column which makes
+    the upsampled tensor asymmetrical on both sides. This does not happen when
+    the upsampling ratio is odd.
+
+    :type input: symbolic 4D tensor
+    :param input: mini-batch of feature map stacks, of shape
+        (batch size, input channels, input rows, input columns).
+        See the optional parameter ``input_shape`` that will be upsampled.
+
+    :type ratio: integer
+    : param ratio: the ratio by which the input is upsampled
+        in the 2D space.
+
+    :type input_shape: None, tuple/list of len 4 of int or
```

Comment:
Errata corrige: let's keep input_shape, which should be [Constant / int / None] * 4